src/core/meshcop/tcat_agent.cpp

```diff
@@ -0,0 +1,519 @@
+    mConnected               = false;
+    mAlreadyCommissioned     = false;
+    mPskdVerified            = false;
+    mPskcVerified            = false;
+    mAppDataReceiveCallback.Clear();
+    mJoinCallback.Clear();
+    LogInfo("TCAT agent stopped");
+}
+
+Error TcatAgent::Connected(MeshCoP::Dtls &aTlsContext)
+{
+    size_t len;
+    Error  error;
+
+    VerifyOrExit(mEnabled, error = kErrorInvalidState);
+    SuccessOrExit(error = aTlsContext.GetThreadAttributeFromPeerCertificate(
+                      3, reinterpret_cast<uint8_t *>(&mCommissionerAuthorizationField), &len,
+                      sizeof(mCommissionerAuthorizationField)));
+    VerifyOrExit(len == sizeof(mCommissionerAuthorizationField), error = kErrorParse);
+    SuccessOrExit(
+        error = aTlsContext.GetThreadAttributeFromOwnCertificate(
+            3, reinterpret_cast<uint8_t *>(&mDeviceAuthorizationField), &len, sizeof(mDeviceAuthorizationField)));
+    VerifyOrExit(len == sizeof(mCommissionerAuthorizationField), error = kErrorParse);
+
+    mCurrentApplicationLayer = OT_TCAT_MESSAGE_TYPE_NONE;
+    mCurrentServiceName[0]   = 0;
+    mConnected               = true;
+    mAlreadyCommissioned     = false; // TODO
+    mPskdVerified            = false;
+    mPskcVerified            = false;
+    LogInfo("TCAT agent connected");
+
+exit:
+    return error;
+}
+
+void TcatAgent::Disconnected()
+{
+    mCurrentApplicationLayer = OT_TCAT_MESSAGE_TYPE_NONE;
+    mConnected               = false;
+    mAlreadyCommissioned     = false;
+    mPskdVerified            = false;
+    mPskcVerified            = false;
+    LogInfo("TCAT agent disconnected");
+}
+
+bool TcatAgent::IsCommandClassAuthorized(CommandClassFlags aCommissionerCommandClassFlags,
+                                         CommandClassFlags aDeviceCommandClassFlags) const
+{
+    bool additionalDeviceRequirementMet = false;
+
+    if ((aCommissionerCommandClassFlags & OT_TCAT_COMMAND_CLASS_FLAG_ACCESS) == 0)
+    {
+        LogInfo("TCAT not authorized");
+        return false;
+    }
+
+    if (aDeviceCommandClassFlags & OT_TCAT_COMMAND_CLASS_FLAG_ACCESS)
+    {
+        additionalDeviceRequirementMet = true;
+    }
+
+    // more to come
+
+    LogInfo("TCAT authorized");
+    return true;
+}
+
+Error TcatAgent::HandleSingleTlv(ot::Message &aIncommingMessage, ot::Message &aOutgoingMessage)
+{
+    Error    error = kErrorParse;
+    ot::Tlv  tlv;
+    uint32_t offset = aIncommingMessage.GetOffset();
+    uint32_t lenght;
+    bool     response = false;
+
+    VerifyOrExit(mConnected, error = kErrorInvalidState);
+    SuccessOrExit(error = aIncommingMessage.Read(offset, tlv));
+
+    if (tlv.IsExtended())
+    {
+        ot::ExtendedTlv extTlv;
+        SuccessOrExit(error = aIncommingMessage.Read(offset, extTlv));
+        lenght = extTlv.GetLength();
+        offset += sizeof(ot::ExtendedTlv);
+    }
+    else
+    {
+        lenght = tlv.GetLength();
+        offset += sizeof(ot::Tlv);
+    }
+
```

Comment:
I think the reason we break the TLVs into groups here is to check what level of access we have.

We can aconsider adding a new method `CanProcessTlv(TlvType aTlvTpe)` that checks this for us based on the `aTlvType` and call it before processing any TLV?

Reply:
This method should be placed in `ExtendedTlv` class?

Reply:
@abtink 
I think I don't understand how this would benefit us.
eg.
```C++
    if(CanProcessTlv(tlv.GetType()))
    {
        if (tlv.GetType() < kTlvSetActiveOperationalDataset)
        {
            error = HandleGeneral(aIncommingMessage, aOutgoingMessage, tlv.GetType(), offset, length, response);
        }
        else if (tlv.GetType() < kTlvGetActiveOperationalDataset)
        {
            error = HandleCommissioning(aIncommingMessage, aOutgoingMessage, tlv.GetType(), offset, length, response);
        }
        else if (tlv.GetType() < kTlvSetLDevIdOperationalCert)
        {
            error = HandleApplication(aIncommingMessage, aOutgoingMessage, tlv.GetType(), offset, length, response);
        }
        else
        {
            error = kErrorInvalidCommand;
        }
     }
```

In this case, parts of code like below would still have to stay in `Handle` methods. Since the boolean method doesn't give us information about what exactly can we handle.
```C++
    if (aTlvType != kTlvSetActiveOperationalDataset && aTlvType != kTlvSetActiveOperationalDatasetAlternative)
    {
        if (!IsCommandClassAuthorized(mCommissionerAuthorizationField.mCommissioningFlags,
                                      mDeviceAuthorizationField.mCommissioningFlags, nullptr))
        {
            error = kErrorRejected;
            ExitNow();
        }
    }
```


I am wondering what I'm missing.

Reply:
It is a suggestion. Basically having one method that checks if we are authrozed to handle certain TLV., followed by a long `switch()` on tlv type which calls `Handle<SpecificTlv>()` to handle each case. I think/hope this make the code simpler?

The main reason I suggested/preferred is to help avoid using numerical comparison on TLV type value to determine this (e.g. `if (tlv.GetType() < kTlvSetActiveOperationalDataset)`, etc). I know this may work nicely as the TLVs are defined today, but it can get tricky to maintain if new TLVs are added in future. It would be easier to have one place that checks what TLV the user is authorized to access.

Reply:
Done. Is this what you had in mind?